### PR TITLE
Speedup binding circuit params

### LINKED ISSUE #5965: Performance regression in parameter binding
### Information

- **Qiskit Terra version**:
- **Python version**:
- **Operating system**:

### What is the current behavior?

Since https://github.com/Qiskit/qiskit-terra/commit/296a745b the nightly benchmarks have flagged a regression in the benchmarks measure parameter binding. This appears quite severe as the number of parameters grow, for example:

https://qiskit.github.io/qiskit/#circuit_construction.ParameterizedCircuitBindBench.time_bind_params?machine=dedicated-benchmarking-softlayer-baremetal&os=Linux%204.15.0-46-generic&ram=16GB&p-width=20&p-gates=2048&p-number%20of%20params=2048&commits=296a745b

which is showing a 69.11x performance regression.

### Steps to reproduce the problem

Bind parameters on a circuit, the benchmark code showing the regression is here:

https://github.com/Qiskit/qiskit/blob/master/test/benchmarks/circuit_construction.py#L88-L102

### What is the expected behavior?

No performance regression

### Suggested solutions

Fix the performance regression

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

Speedup binding and returning circuit parameters by

* only sorting parameters when it's really needed 
* caching the sorted parameters

This should tackle the ~70x slowdown as presented in https://qiskit.github.io/qiskit/#circuit_construction.ParameterizedCircuitBindBench.time_bind_params?machine=dedicated-benchmarking-softlayer-baremetal&os=Linux%204.15.0-46-generic&ram=16GB&p-width=20&p-gates=2048&p-number%20of%20params=2048&commits=296a745b.

### Details and comments

On a 1000 parameter circuit binding 10 times currently takes
```
dict took 74.88096404075623s # bind via dict
array took 74.05472993850708s  # bind via array
```
with this PR its 
```
dict took 1.5851850509643555s 
array took 0.6309740543365479s
``` 

Fixes #5965

### CODE REVIEW

**Review by ewinston — COMMENTED** [review]
Over all looks good. Since some of the speed ups come from catching calls to e.g. `len(circ.parameters)` or membership such as `a in circ.parameters`, I wonder if some of this logic can be moved into ParamterTable or ParameterView with `__len__` and `__contains__` implemented?

File: qiskit/circuit/quantumcircuit.py [hunk 1]
```diff
@@ -1027,6 +1031,9 @@ def _update_parameter_table(self, instruction):
                                 'Name conflict on adding parameter: {}'.format(parameter.name))
                         self._parameter_table[parameter] = [(instruction, param_index)]
 
+                        # clear cache if new parameter is added
+                        self._parameters = None
```
[the remark this answers is not in the document]
Comment (Cryoris) [reply qiskit/circuit/quantumcircuit.py:1035]: There are several places where the caching can be further optimized, e.g. also in deletion. I had that implemented before (using a BTree) but that adds logarithmic overhead to adding parameterized gates or might have other performance drawbacks. So instead I went with just a very simple caching 🙂 
But I'm happy to improve the caching, but I'd prefer to do this in a follow-up, if possible

**Comment by Cryoris** [comment]
It would be nice to have a more uniform interface for accessing the parameters that would also allow `__len__` and `__contains__`, I agree! @molar-volume also suggested that while working on #5961.

It's not that straightforward though, since 
1. the parameter table doesn't know all parameters (global phase might be parameterized) 
2. the parameter view object was designed to be temporary and is currently only constructed upon returning the parameters
If we instead keep a `ParameterView` object internally (that also might do more efficient caching like you suggested in the comments), we could improve all these points. But this is likely better done in a follow-up, too 🙂

File: qiskit/circuit/quantumcircuit.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ewinston) [reply qiskit/circuit/quantumcircuit.py:1046]: That sounds fine.

**Review by ewinston — APPROVED** [review]

**Comment by kdk** [comment]
> It would be nice to have a more uniform interface for accessing the parameters that would also allow `__len__` and `__contains__`, I agree! @molar-volume also suggested that while working on #5961.

@Cryoris is there an issue to track these discussions so we can follow up once these PRs are merged? Can you open one if not?

**Comment by molar-volume** [comment]
I opened a draft suggesting changes to `ParameterView` and exploring options for simplifying `ParameterTable` as well https://github.com/Qiskit/qiskit-terra/pull/6074

**Pull request merged**